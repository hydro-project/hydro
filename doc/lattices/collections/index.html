<!DOCTYPE html><html lang="en"><head><meta charset="utf-8"><meta name="viewport" content="width=device-width, initial-scale=1.0"><meta name="generator" content="rustdoc"><meta name="description" content="Simple singleton or array collection with `cc_traits` implementations."><title>lattices::collections - Rust</title><script>if(window.location.protocol!=="file:")document.head.insertAdjacentHTML("beforeend","SourceSerif4-Regular-6b053e98.ttf.woff2,FiraSans-Italic-81dc35de.woff2,FiraSans-Regular-0fe48ade.woff2,FiraSans-MediumItalic-ccf7e434.woff2,FiraSans-Medium-e1aa3f0a.woff2,SourceCodePro-Regular-8badfe75.ttf.woff2,SourceCodePro-Semibold-aa29a496.ttf.woff2".split(",").map(f=>`<link rel="preload" as="font" type="font/woff2" crossorigin href="../../static.files/${f}">`).join(""))</script><link rel="stylesheet" href="../../static.files/normalize-9960930a.css"><link rel="stylesheet" href="../../static.files/rustdoc-20739d33.css"><meta name="rustdoc-vars" data-root-path="../../" data-static-root-path="../../static.files/" data-current-crate="lattices" data-themes="" data-resource-suffix="" data-rustdoc-version="1.88.0-nightly (10fa3c449 2025-04-26)" data-channel="nightly" data-search-js="search-f7877310.js" data-settings-js="settings-5514c975.js" ><script src="../../static.files/storage-4e99c027.js"></script><script defer src="../sidebar-items.js"></script><script defer src="../../static.files/main-7ef8a74a.js"></script><noscript><link rel="stylesheet" href="../../static.files/noscript-893ab5e7.css"></noscript><link rel="alternate icon" type="image/png" href="../../static.files/favicon-32x32-6580c154.png"><link rel="icon" type="image/svg+xml" href="../../static.files/favicon-044be391.svg"></head><body class="rustdoc mod"><!--[if lte IE 11]><div class="warning">This old browser is unsupported and will most likely display funky things.</div><![endif]--><nav class="mobile-topbar"><button class="sidebar-menu-toggle" title="show sidebar"></button></nav><nav class="sidebar"><div class="sidebar-crate"><h2><a href="../../lattices/index.html">lattices</a><span class="version">0.6.1</span></h2></div><div class="sidebar-elems"><section id="rustdoc-toc"><h2 class="location"><a href="#">Module collections</a></h2><h3><a href="#structs">Module Items</a></h3><ul class="block"><li><a href="#structs" title="Structs">Structs</a></li><li><a href="#traits" title="Traits">Traits</a></li></ul></section><div id="rustdoc-modnav"><h2 class="in-crate"><a href="../index.html">In crate lattices</a></h2></div></div></nav><div class="sidebar-resizer"></div><main><div class="width-limiter"><rustdoc-search></rustdoc-search><section id="main-content" class="content"><div class="main-heading"><div class="rustdoc-breadcrumbs"><a href="../index.html">lattices</a></div><h1>Module <span>collections</span><button id="copy-path" title="Copy item path to clipboard">Copy item path</button></h1><rustdoc-toolbar></rustdoc-toolbar><span class="sub-heading"><a class="src" href="../../src/lattices/collections.rs.html#1-1203">Source</a> </span></div><details class="toggle top-doc" open><summary class="hideme"><span>Expand description</span></summary><div class="docblock"><p>Simple singleton or array collection with [<code>cc_traits</code>] implementations.</p>
</div></details><h2 id="structs" class="section-header">Structs<a href="#structs" class="anchor">ยง</a></h2><dl class="item-table"><dt><a class="struct" href="struct.ArrayMap.html" title="struct lattices::collections::ArrayMap">Array<wbr>Map</a></dt><dd>An array wrapper representing a fixed-size map.</dd><dt><a class="struct" href="struct.ArraySet.html" title="struct lattices::collections::ArraySet">Array<wbr>Set</a></dt><dd>An array wrapper representing a fixed-size set (modulo duplicate items).</dd><dt><a class="struct" href="struct.EmptyMap.html" title="struct lattices::collections::EmptyMap">Empty<wbr>Map</a></dt><dd>A key-value entry wrapper representing a singleton map.</dd><dt><a class="struct" href="struct.EmptySet.html" title="struct lattices::collections::EmptySet">Empty<wbr>Set</a></dt><dd>A type that will always be an empty set.</dd><dt><a class="struct" href="struct.OptionMap.html" title="struct lattices::collections::OptionMap">Option<wbr>Map</a></dt><dd>A key-value entry wrapper around <code>Option&lt;(K, V)&gt;</code> representing a singleton or empty map.</dd><dt><a class="struct" href="struct.OptionSet.html" title="struct lattices::collections::OptionSet">Option<wbr>Set</a></dt><dd>A wrapper around <code>Option</code>, representing either a singleton or empty set.</dd><dt><a class="struct" href="struct.SingletonMap.html" title="struct lattices::collections::SingletonMap">Singleton<wbr>Map</a></dt><dd>A key-value entry wrapper representing a singleton map.</dd><dt><a class="struct" href="struct.SingletonSet.html" title="struct lattices::collections::SingletonSet">Singleton<wbr>Set</a></dt><dd>A wrapper around an item, representing a singleton set.</dd><dt><a class="struct" href="struct.VecMap.html" title="struct lattices::collections::VecMap">VecMap</a></dt><dd>A <a href="https://doc.rust-lang.org/nightly/alloc/vec/struct.Vec.html" title="struct alloc::vec::Vec"><code>Vec</code></a>-wrapper representing a naively implemented map.</dd><dt><a class="struct" href="struct.VecSet.html" title="struct lattices::collections::VecSet">VecSet</a></dt><dd>A <a href="https://doc.rust-lang.org/nightly/alloc/vec/struct.Vec.html" title="struct alloc::vec::Vec"><code>Vec</code></a>-wrapper representing a naively-implemented set.</dd></dl><h2 id="traits" class="section-header">Traits<a href="#traits" class="anchor">ยง</a></h2><dl class="item-table"><dt><a class="trait" href="trait.MapMapValues.html" title="trait lattices::collections::MapMapValues">MapMap<wbr>Values</a></dt><dd>Trait for transforming the values of a map without changing the overall type of the data structure.</dd></dl></section></div></main></body></html>