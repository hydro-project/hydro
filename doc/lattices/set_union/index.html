<!DOCTYPE html><html lang="en"><head><meta charset="utf-8"><meta name="viewport" content="width=device-width, initial-scale=1.0"><meta name="generator" content="rustdoc"><meta name="description" content="Module containing the `SetUnion` lattice and aliases for different datastructures."><title>lattices::set_union - Rust</title><script>if(window.location.protocol!=="file:")document.head.insertAdjacentHTML("beforeend","SourceSerif4-Regular-6b053e98.ttf.woff2,FiraSans-Italic-81dc35de.woff2,FiraSans-Regular-0fe48ade.woff2,FiraSans-MediumItalic-ccf7e434.woff2,FiraSans-Medium-e1aa3f0a.woff2,SourceCodePro-Regular-8badfe75.ttf.woff2,SourceCodePro-Semibold-aa29a496.ttf.woff2".split(",").map(f=>`<link rel="preload" as="font" type="font/woff2" crossorigin href="../../static.files/${f}">`).join(""))</script><link rel="stylesheet" href="../../static.files/normalize-9960930a.css"><link rel="stylesheet" href="../../static.files/rustdoc-20739d33.css"><meta name="rustdoc-vars" data-root-path="../../" data-static-root-path="../../static.files/" data-current-crate="lattices" data-themes="" data-resource-suffix="" data-rustdoc-version="1.88.0-nightly (10fa3c449 2025-04-26)" data-channel="nightly" data-search-js="search-f7877310.js" data-settings-js="settings-5514c975.js" ><script src="../../static.files/storage-4e99c027.js"></script><script defer src="../sidebar-items.js"></script><script defer src="../../static.files/main-7ef8a74a.js"></script><noscript><link rel="stylesheet" href="../../static.files/noscript-893ab5e7.css"></noscript><link rel="alternate icon" type="image/png" href="../../static.files/favicon-32x32-6580c154.png"><link rel="icon" type="image/svg+xml" href="../../static.files/favicon-044be391.svg"></head><body class="rustdoc mod"><!--[if lte IE 11]><div class="warning">This old browser is unsupported and will most likely display funky things.</div><![endif]--><nav class="mobile-topbar"><button class="sidebar-menu-toggle" title="show sidebar"></button></nav><nav class="sidebar"><div class="sidebar-crate"><h2><a href="../../lattices/index.html">lattices</a><span class="version">0.6.1</span></h2></div><div class="sidebar-elems"><section id="rustdoc-toc"><h2 class="location"><a href="#">Module set_<wbr>union</a></h2><h3><a href="#structs">Module Items</a></h3><ul class="block"><li><a href="#structs" title="Structs">Structs</a></li><li><a href="#types" title="Type Aliases">Type Aliases</a></li></ul></section><div id="rustdoc-modnav"><h2 class="in-crate"><a href="../index.html">In crate lattices</a></h2></div></div></nav><div class="sidebar-resizer"></div><main><div class="width-limiter"><rustdoc-search></rustdoc-search><section id="main-content" class="content"><div class="main-heading"><div class="rustdoc-breadcrumbs"><a href="../index.html">lattices</a></div><h1>Module <span>set_union</span><button id="copy-path" title="Copy item path to clipboard">Copy item path</button></h1><rustdoc-toolbar></rustdoc-toolbar><span class="sub-heading"><a class="src" href="../../src/lattices/set_union.rs.html#1-312">Source</a> </span></div><details class="toggle top-doc" open><summary class="hideme"><span>Expand description</span></summary><div class="docblock"><p>Module containing the <a href="struct.SetUnion.html" title="struct lattices::set_union::SetUnion"><code>SetUnion</code></a> lattice and aliases for different datastructures.</p>
</div></details><h2 id="structs" class="section-header">Structs<a href="#structs" class="anchor">ยง</a></h2><dl class="item-table"><dt><a class="struct" href="struct.CartesianProductBimorphism.html" title="struct lattices::set_union::CartesianProductBimorphism">Cartesian<wbr>Product<wbr>Bimorphism</a></dt><dd>Bimorphism for the cartesian product of two sets. Output is a set of all possible pairs of
items from the two input sets.</dd><dt><a class="struct" href="struct.SetUnion.html" title="struct lattices::set_union::SetUnion">SetUnion</a></dt><dd>Set-union lattice.</dd></dl><h2 id="types" class="section-header">Type Aliases<a href="#types" class="anchor">ยง</a></h2><dl class="item-table"><dt><a class="type" href="type.SetUnionArray.html" title="type lattices::set_union::SetUnionArray">SetUnion<wbr>Array</a></dt><dd><a href="../collections/struct.ArraySet.html" title="struct lattices::collections::ArraySet"><code>crate::collections::ArraySet</code></a>-backed <a href="struct.SetUnion.html" title="struct lattices::set_union::SetUnion"><code>SetUnion</code></a> lattice.</dd><dt><a class="type" href="type.SetUnionBTreeSet.html" title="type lattices::set_union::SetUnionBTreeSet">SetUnionB<wbr>Tree<wbr>Set</a></dt><dd><a href="https://doc.rust-lang.org/nightly/alloc/collections/btree/set/struct.BTreeSet.html" title="struct alloc::collections::btree::set::BTreeSet"><code>std::collections::BTreeSet</code></a>-backed <a href="struct.SetUnion.html" title="struct lattices::set_union::SetUnion"><code>SetUnion</code></a> lattice.</dd><dt><a class="type" href="type.SetUnionHashSet.html" title="type lattices::set_union::SetUnionHashSet">SetUnion<wbr>Hash<wbr>Set</a></dt><dd><a href="https://doc.rust-lang.org/nightly/std/collections/hash/set/struct.HashSet.html" title="struct std::collections::hash::set::HashSet"><code>std::collections::HashSet</code></a>-backed <a href="struct.SetUnion.html" title="struct lattices::set_union::SetUnion"><code>SetUnion</code></a> lattice.</dd><dt><a class="type" href="type.SetUnionOptionSet.html" title="type lattices::set_union::SetUnionOptionSet">SetUnion<wbr>Option<wbr>Set</a></dt><dd><a href="https://doc.rust-lang.org/nightly/core/option/enum.Option.html" title="enum core::option::Option"><code>Option</code></a>-backed <a href="struct.SetUnion.html" title="struct lattices::set_union::SetUnion"><code>SetUnion</code></a> lattice.</dd><dt><a class="type" href="type.SetUnionSingletonSet.html" title="type lattices::set_union::SetUnionSingletonSet">SetUnion<wbr>Singleton<wbr>Set</a></dt><dd><a href="../collections/struct.SingletonSet.html" title="struct lattices::collections::SingletonSet"><code>crate::collections::SingletonSet</code></a>-backed <a href="struct.SetUnion.html" title="struct lattices::set_union::SetUnion"><code>SetUnion</code></a> lattice.</dd><dt><a class="type" href="type.SetUnionVec.html" title="type lattices::set_union::SetUnionVec">SetUnion<wbr>Vec</a></dt><dd><a href="https://doc.rust-lang.org/nightly/alloc/vec/struct.Vec.html" title="struct alloc::vec::Vec"><code>Vec</code></a>-backed <a href="struct.SetUnion.html" title="struct lattices::set_union::SetUnion"><code>SetUnion</code></a> lattice.</dd></dl></section></div></main></body></html>