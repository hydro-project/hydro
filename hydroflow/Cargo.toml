[package]
name = "hydroflow"
version = "0.1.0"
edition = "2021"

[features]
default = [ "async", "macros" ]
async = [ "futures" ]
macros = [ "hydroflow_macro", "hydroflow_datalog" ]

[[example]]
name = "echoserver"

[[example]]
name = "echo_serde_json"

[[example]]
name = "chat"

[[example]]
name = "graph_reachability"

[[example]]
name = "graph_unreachability"

[[example]]
name = "three_clique"

[[example]]
name = "two_pc"

[[example]]
name = "deadlock_detector"

[[example]]
name = "kvs"

[dependencies]
bincode = "1.3"
byteorder = "1.4.3"
bytes = "1.1.0"
futures = { version = "0.3", optional = true }
hydroflow_datalog = { optional = true, path = "../hydroflow_datalog" }
hydroflow_lang = { path = "../hydroflow_lang" }
hydroflow_macro = { optional = true, path = "../hydroflow_macro" }
pin-project-lite = "0.2.9"
pusherator = { path = "../pusherator" }
ref-cast = "1.0"
regex = "1"
sealed = "0.4"
serde = { version = "1", features = [ "derive" ] }
serde_json = "1"
slotmap = "1.0"
static_assertions = "1.1.0"
tokio = { version = "1.16", features = [ "full" ] }
tokio-stream = { version = "0.1.10", features = [ "io-util" ] }
tokio-util = { version = "0.7.4", features = [ "net", "codec" ] }
variadics = { path = "../variadics" }

[dev-dependencies]
chrono = { version = "0.4.20", features = [ "serde" ], default-features = false }
clap = { version = "3.0.14", features = [ "derive" ] }
colored = "2.0"
criterion = { version = "0.3", features = [ "async_tokio" ] }
futures = { version = "0.3" }
itertools = "0.10.3"
rand = "0.8.4"
textnonce = "1.0.0"
time = "0.3"
trybuild = "1.0.71"
hdrhistogram = "7"
zipf = "7.0.0"
core_affinity = "0.5.10"
