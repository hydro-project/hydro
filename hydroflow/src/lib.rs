#![feature(never_type)]
#![feature(type_alias_impl_trait)]
#![allow(type_alias_bounds)]
#![allow(clippy::let_and_return)]
#![allow(clippy::iter_with_drain)]
#![allow(clippy::explicit_auto_deref)]

pub mod builder;
pub mod compiled;
pub mod lang;
pub mod props;
pub mod scheduled;

pub use hydroflow_macro::*;

pub use pusherator;
pub use tokio;
pub use tuple_list::tuple_list as tl;
pub use tuple_list::tuple_list_type as tt;

#[cfg(doctest)]
mod booktest {
    macro_rules! booktest {
        ($i:ident) => {
            #[doc = include_str!(concat!("../../book/", stringify!($i), ".md"))]
            mod $i {}
        };
    }
    booktest!(example_1);
    booktest!(example_2);
    booktest!(example_3);
    booktest!(example_4);
    booktest!(example_5);

    booktest!(example_1_surface);
    booktest!(example_2_surface);
    booktest!(example_3_surface);
    booktest!(example_4_surface);
    booktest!(example_5_surface);

    booktest!(surface_syntax);
    booktest!(surface_embedding);
    booktest!(surface_flows);
    booktest!(surface_data);
    booktest!(surface_ops);
}
