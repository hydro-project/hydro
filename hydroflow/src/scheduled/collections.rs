pub struct Single<T>(pub T);

pub struct Iter<I: Iterator>(pub I);
