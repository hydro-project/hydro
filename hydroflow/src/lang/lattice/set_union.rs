use std::cmp::Ordering;
use std::iter::FromIterator;

use super::{Compare, Convert, Debottom, Lattice, LatticeRepr, Merge};

use crate::lang::collections::Collection;
use crate::lang::tag;

pub struct SetUnion<T> {
    _phantom: std::marker::PhantomData<T>,
}
impl<T> Lattice for SetUnion<T> {}

pub trait SetTag<T>: tag::Tag1<T> {}
impl<T> SetTag<T> for tag::HASH_SET {}
impl<T> SetTag<T> for tag::BTREE_SET {}
impl<T> SetTag<T> for tag::VEC {}
impl<T> SetTag<T> for tag::SINGLE {}
impl<T> SetTag<T> for tag::OPTION {}
impl<T, const N: usize> SetTag<T> for tag::ARRAY<N> {}
impl<T, const N: usize> SetTag<T> for tag::MASKED_ARRAY<N> {}

pub struct SetUnionRepr<Tag: SetTag<T>, T> {
    _phantom: std::marker::PhantomData<(Tag, T)>,
}

impl<Tag: SetTag<T>, T> LatticeRepr for SetUnionRepr<Tag, T>
where
    Tag::Bind: Clone,
{
    type Lattice = SetUnion<T>;
    type Repr = Tag::Bind;
}

impl<T, SelfTag: SetTag<T>, DeltaTag: SetTag<T>> Merge<SetUnionRepr<DeltaTag, T>>
    for SetUnionRepr<SelfTag, T>
where
    SetUnionRepr<SelfTag, T>: LatticeRepr<Lattice = SetUnion<T>>,
    SetUnionRepr<DeltaTag, T>: LatticeRepr<Lattice = SetUnion<T>>,
    <SetUnionRepr<SelfTag, T> as LatticeRepr>::Repr: Collection<T, ()> + Extend<T>,
    <SetUnionRepr<DeltaTag, T> as LatticeRepr>::Repr: IntoIterator<Item = T>,
{
    fn merge(
        this: &mut <SetUnionRepr<SelfTag, T> as LatticeRepr>::Repr,
        delta: <SetUnionRepr<DeltaTag, T> as LatticeRepr>::Repr,
    ) -> bool {
        let old_len = this.len();
        this.extend(delta);
        this.len() > old_len
    }
}

impl<T, SelfTag: SetTag<T>, TargetTag: SetTag<T>> Convert<SetUnionRepr<TargetTag, T>>
    for SetUnionRepr<SelfTag, T>
where
    SetUnionRepr<SelfTag, T>: LatticeRepr<Lattice = SetUnion<T>>,
    SetUnionRepr<TargetTag, T>: LatticeRepr<Lattice = SetUnion<T>>,
    <SetUnionRepr<SelfTag, T> as LatticeRepr>::Repr: IntoIterator<Item = T>,
    <SetUnionRepr<TargetTag, T> as LatticeRepr>::Repr: FromIterator<T>,
{
    fn convert(
        this: <SetUnionRepr<SelfTag, T> as LatticeRepr>::Repr,
    ) -> <SetUnionRepr<TargetTag, T> as LatticeRepr>::Repr {
        this.into_iter().collect()
    }
}

impl<T: 'static, SelfTag: SetTag<T>, TargetTag: SetTag<T>> Compare<SetUnionRepr<TargetTag, T>>
    for SetUnionRepr<SelfTag, T>
where
    SetUnionRepr<SelfTag, T>: LatticeRepr<Lattice = SetUnion<T>>,
    SetUnionRepr<TargetTag, T>: LatticeRepr<Lattice = SetUnion<T>>,
    <SetUnionRepr<SelfTag, T> as LatticeRepr>::Repr: Collection<T, ()>,
    <SetUnionRepr<TargetTag, T> as LatticeRepr>::Repr: Collection<T, ()>,
{
    fn compare(
        this: &<SetUnionRepr<SelfTag, T> as LatticeRepr>::Repr,
        other: &<SetUnionRepr<TargetTag, T> as LatticeRepr>::Repr,
    ) -> Option<Ordering> {
        match this.len().cmp(&other.len()) {
            Ordering::Greater => {
                if other.keys().all(|key| this.get(key).is_some()) {
                    Some(Ordering::Greater)
                } else {
                    None
                }
            }
            Ordering::Equal => {
                if this.keys().all(|key| other.get(key).is_some()) {
                    Some(Ordering::Equal)
                } else {
                    None
                }
            }
            Ordering::Less => {
                if this.keys().all(|key| other.get(key).is_some()) {
                    Some(Ordering::Less)
                } else {
                    None
                }
            }
        }
    }
}

// impl<Tag: SetTag<T>, T> Debottom for SetUnionRepr<Tag, T>
// where
//     Tag::Bind: Clone,
//     Self::Repr: Collection<T, ()>,
// {
//     fn is_bottom(this: &Self::Repr) -> bool {
//         this.is_empty()
//     }
// }
impl<T: Clone> Debottom for SetUnionRepr<tag::OPTION, T> {
    fn is_bottom(this: &Self::Repr) -> bool {
        this.is_none()
    }

    type DebottomLr = SetUnionRepr<tag::SINGLE, T>;
    fn debottom(this: Self::Repr) -> Option<<Self::DebottomLr as LatticeRepr>::Repr> {
        this.map(crate::lang::collections::Single)
    }
}

#[cfg(test)]
fn __assert_merges() {
    use static_assertions::{assert_impl_all, assert_not_impl_any};

    assert_impl_all!(SetUnionRepr<tag::HASH_SET, u32>:
        Merge<SetUnionRepr<tag::HASH_SET, u32>>,
        Merge<SetUnionRepr<tag::BTREE_SET, u32>>,
        Merge<SetUnionRepr<tag::VEC, u32>>,
        Merge<SetUnionRepr<tag::SINGLE, u32>>,
        Merge<SetUnionRepr<tag::OPTION, u32>>,
        Merge<SetUnionRepr<tag::ARRAY<8>, u32>>,
        Merge<SetUnionRepr<tag::MASKED_ARRAY<8>, u32>>,
    );

    assert_impl_all!(SetUnionRepr<tag::BTREE_SET, u32>:
        Merge<SetUnionRepr<tag::HASH_SET, u32>>,
        Merge<SetUnionRepr<tag::BTREE_SET, u32>>,
        Merge<SetUnionRepr<tag::VEC, u32>>,
        Merge<SetUnionRepr<tag::SINGLE, u32>>,
        Merge<SetUnionRepr<tag::OPTION, u32>>,
        Merge<SetUnionRepr<tag::ARRAY<8>, u32>>,
        Merge<SetUnionRepr<tag::MASKED_ARRAY<8>, u32>>,
    );

    assert_impl_all!(SetUnionRepr<tag::VEC, u32>:
        Merge<SetUnionRepr<tag::HASH_SET, u32>>,
        Merge<SetUnionRepr<tag::BTREE_SET, u32>>,
        Merge<SetUnionRepr<tag::VEC, u32>>,
        Merge<SetUnionRepr<tag::SINGLE, u32>>,
        Merge<SetUnionRepr<tag::OPTION, u32>>,
        Merge<SetUnionRepr<tag::ARRAY<8>, u32>>,
        Merge<SetUnionRepr<tag::MASKED_ARRAY<8>, u32>>,
    );

    assert_not_impl_any!(SetUnionRepr<tag::MASKED_ARRAY<8>, u32>:
        Merge<SetUnionRepr<tag::HASH_SET, u32>>,
        Merge<SetUnionRepr<tag::BTREE_SET, u32>>,
        Merge<SetUnionRepr<tag::VEC, u32>>,
        Merge<SetUnionRepr<tag::SINGLE, u32>>,
        Merge<SetUnionRepr<tag::OPTION, u32>>,
        Merge<SetUnionRepr<tag::ARRAY<8>, u32>>,
        Merge<SetUnionRepr<tag::MASKED_ARRAY<8>, u32>>,
    );
}
