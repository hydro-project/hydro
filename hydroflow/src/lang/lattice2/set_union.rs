//! Set-union compound lattice.
//!
//! Merging set-union lattices is done by unioning the keys.

use std::cmp::Ordering;

use crate::lang::{collections::Collection, tag};

use super::{Compare, ConvertFrom, Merge};

/// A set-union lattice.
///
/// `Tag` specifies what datastructure to use, allowing us to deal with different datastructures
/// generically.
#[repr(transparent)]
pub struct SetUnion<Tag, T>(Tag::Bind)
where
    Tag: tag::Tag1<T>;
impl<Tag, T> SetUnion<Tag, T>
where
    Tag: tag::Tag1<T>,
{
    /// Create a new `SetUnion` from a set.
    pub fn new(val: impl Into<Tag::Bind>) -> Self {
        Self(val.into())
    }
}

impl<TagSelf, TagOther, T> Merge<SetUnion<TagOther, T>> for SetUnion<TagSelf, T>
where
    TagSelf: tag::Tag1<T>,
    TagOther: tag::Tag1<T>,
    TagSelf::Bind: Collection<T, ()> + Extend<T>,
    TagOther::Bind: IntoIterator<Item = T>,
{
    fn merge(&mut self, other: SetUnion<TagOther, T>) -> bool {
        let old_len = self.0.len();
        self.0.extend(other.0);
        self.0.len() > old_len
    }
}

impl<TagSelf, TagOther, T> ConvertFrom<SetUnion<TagOther, T>> for SetUnion<TagSelf, T>
where
    TagSelf: tag::Tag1<T>,
    TagOther: tag::Tag1<T>,
    TagSelf::Bind: FromIterator<T>,
    TagOther::Bind: Collection<T, ()>,
{
    fn from(other: SetUnion<TagOther, T>) -> Self {
        Self(
            other
                .0
                .into_entries()
                .map(|(t_other, ())| t_other)
                .collect(),
        )
    }
}

impl<Tag, T> Default for SetUnion<Tag, T>
where
    Tag: tag::Tag1<T>,
    Tag::Bind: Default,
{
    fn default() -> Self {
        Self(Default::default())
    }
}

impl<TagSelf, TagOther, T> Compare<SetUnion<TagOther, T>> for SetUnion<TagSelf, T>
where
    TagSelf: tag::Tag1<T>,
    TagOther: tag::Tag1<T>,
    TagSelf::Bind: Collection<T, ()>,
    TagOther::Bind: Collection<T, ()>,
{
    fn compare(&self, other: &SetUnion<TagOther, T>) -> Option<Ordering> {
        match self.0.len().cmp(&other.0.len()) {
            Ordering::Greater => {
                if self.0.keys().all(|key| other.0.get(key).is_some()) {
                    Some(Ordering::Greater)
                } else {
                    None
                }
            }
            Ordering::Equal => {
                if self.0.keys().all(|key| other.0.get(key).is_some()) {
                    Some(Ordering::Equal)
                } else {
                    None
                }
            }
            Ordering::Less => {
                if other.0.keys().all(|key| self.0.get(key).is_some()) {
                    Some(Ordering::Less)
                } else {
                    None
                }
            }
        }
    }
}

/// [`std::collections::HashSet`]-backed [`SetUnion`] lattice.
pub type SetUnionHashSet<T> = SetUnion<tag::HASH_SET, T>;

/// [`std::collections::BTreeSet`]-backed [`SetUnion`] lattice.
pub type SetUnionBTreeSet<T> = SetUnion<tag::BTREE_SET, T>;

/// [`Vec`]-backed [`SetUnion`] lattice.
pub type SetUnionVec<T> = SetUnion<tag::VEC, T>;

/// Array-backed [`SetUnion`] lattice.
pub type SetUnionArray<T, const N: usize> = SetUnion<tag::ARRAY<N>, T>;

/// [`crate::lang::collections::MaskedArray`]-backed [`SetUnion`] lattice.
pub type SetUnionMaskedArray<T, const N: usize> = SetUnion<tag::MASKED_ARRAY<N>, T>;

/// [`crate::lang::collections::Single`]-backed [`SetUnion`] lattice.
pub type SetUnionSingle<T> = SetUnion<tag::SINGLE, T>;

/// [`Option`]-backed [`SetUnion`] lattice.
pub type SetUnionOption<T> = SetUnion<tag::OPTION, T>;

#[cfg(test)]
mod test {
    use super::*;

    use crate::lang::collections::Single;

    #[test]
    fn test_set_union() {
        let mut my_set_a = SetUnion::<tag::HASH_SET, &str>(Default::default());
        let my_set_b = SetUnion::<tag::BTREE_SET, &str>(Default::default());
        let my_set_c = SetUnion::<tag::SINGLE, _>(Single("hello world"));
        my_set_a.merge(my_set_b);
        my_set_a.merge(my_set_c);
    }
}
