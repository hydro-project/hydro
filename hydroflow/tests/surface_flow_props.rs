use std::error::Error;

use hydroflow::lattices::collections::SingletonSet;
use hydroflow::lattices::set_union::{SetUnion, SetUnionHashSet, SetUnionSingletonSet};
use hydroflow::{assert_graphvis_snapshots, hydroflow_syntax};

#[test]
pub fn test_basic() -> Result<(), Box<dyn Error>> {
    let mut hf = hydroflow_syntax! {
        my_tee = source_iter_delta((0..10).map(SetUnionSingletonSet::new_from))
            -> map(|SetUnion(SingletonSet(x))| SetUnion(SingletonSet(x + 5)))
            -> tee();

        my_tee
            -> cast(None)
            -> map(|SetUnion(SingletonSet(x))| 10 * x)
            -> for_each(|x| println!("seq {:?}", x));

        my_tee
            -> for_each(|s| println!("delta {:?}", s));

        my_tee
            -> map(|SetUnion(SingletonSet(x))| SetUnionHashSet::new_from([x]))
            -> lattice_reduce::<'static, SetUnionHashSet<_>>()
            -> for_each(|s| println!("cumul {:?}", s));
    };
    hf.run_available();

    assert_graphvis_snapshots!(hf);

    Ok(())
}
