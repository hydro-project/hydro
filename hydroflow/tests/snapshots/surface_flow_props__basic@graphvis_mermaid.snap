---
source: hydroflow/tests/surface_flow_props.rs
expression: hf.meta_graph().unwrap().to_mermaid()
---
%%{init:{'theme':'base','themeVariables':{'clusterBkg':'#ddd','clusterBorder':'#888'}}}%%
flowchart TD
classDef pullClass fill:#8af,stroke:#000,text-align:left,white-space:pre
classDef pushClass fill:#ff8,stroke:#000,text-align:left,white-space:pre
linkStyle default stroke:#aaa,stroke-width:4px,color:red,font-size:1.5em;
subgraph sg_1v1 ["sg_1v1 stratum 0"]
    1v1[\"(1v1) <code>source_iter_delta((0..10).map(SetUnionSingletonSet::new_from))</code>"/]:::pullClass
    2v1[\"(2v1) <code>map(|SetUnion(SingletonSet(x))| SetUnion(SingletonSet(x + 5)))</code>"/]:::pullClass
    3v1[/"(3v1) <code>tee()</code>"\]:::pushClass
    4v1[/"(4v1) <code>cast(None)</code>"\]:::pushClass
    5v1[/"(5v1) <code>map(|SetUnion(SingletonSet(x))| 10 * x)</code>"\]:::pushClass
    6v1[/"(6v1) <code>for_each(|x| println!(&quot;seq {:?}&quot;, x))</code>"\]:::pushClass
    7v1[/"(7v1) <code>for_each(|s| println!(&quot;delta {:?}&quot;, s))</code>"\]:::pushClass
    1v1-.->2v1
    2v1-.->3v1
    3v1-.->4v1
    3v1-.->7v1
    4v1-->5v1
    5v1-->6v1
    subgraph sg_1v1_var_my_tee ["var <tt>my_tee</tt>"]
        1v1
        2v1
        3v1
    end
end

